src/auxiliary/mpi_arrays.jl

```diff
@@ -0,0 +1,152 @@
+  mpi_rank = MPI.Comm_rank(MPI.COMM_WORLD)
+  TrixiMPIArray{T, N, typeof(u_local)}(u_local, mpi_comm, mpi_rank)
+end
+
+
+# TODO: MPI. Adapt
+# - wrap_array - done
+# - wrap_array_native - should not be changed since it should return a plain `Array`
+# - return type of initialization stuff when setting an IC with MPI - handled
+#   by allocate_coefficients, done
+# - dispatch on this array type instead of parallel trees etc. and use
+#   `parent(u)` to get local versions instead of `invoke`
+#   TODO: MPI. When do we want to dispatch on `u::TrixiMPIArray` and when on
+#              an MPI parallel mesh type? Right now, we dispatch a lot on
+#              something like `mesh::ParallelTreeMesh{2}`, which can be nice
+#              since it is consistent with dispatch on `mesh::TreeMesh{2}`
+#              in the general case. At first, I just dispatched on TrixiMPIArray
+#              a few times, in particular when using `parent` instead of `invoke`
+#              machinery.
+
+
+# Custom interface and general Base interface not covered by other parts below
+Base.parent(u::TrixiMPIArray) = u.u_local
+Base.resize!(u::TrixiMPIArray, new_size) = resize!(parent(u), new_size)
+
+mpi_comm(u::TrixiMPIArray) = u.mpi_comm
+mpi_rank(u::TrixiMPIArray) = u.mpi_rank
+# TODO: MPI. What about the following interface functions?
+# mpi_nranks(u::TrixiMPIArray) = MPI_SIZE[]
+# mpi_isparallel(u::TrixiMPIArray) = MPI_IS_PARALLEL[]
+
+
+# Implementation of the abstract array interface of Base
+# See https://docs.julialang.org/en/v1/manual/interfaces/#man-interface-array
+Base.size(u::TrixiMPIArray) = size(parent(u))
+Base.getindex(u::TrixiMPIArray, idx) = getindex(parent(u), idx)
+Base.setindex!(u::TrixiMPIArray, v, idx) = setindex!(parent(u), v, idx)
+Base.IndexStyle(::Type{TrixiMPIArray{T, N, Parent}}) where {T, N, Parent} = IndexStyle(Parent)
+Base.similar(u::TrixiMPIArray, ::Type{S}, dims::NTuple{N, Int}) where {S, N}	= TrixiMPIArray(similar(parent(u), S, dims))
+Base.axes(u::TrixiMPIArray)	= axes(parent(u))
+
+
+# Implementation of the strided array interface of Base
+# See https://docs.julialang.org/en/v1/manual/interfaces/#man-interface-strided-arrays
+Base.strides(u::TrixiMPIArray) = strides(parent(u))
+Base.unsafe_convert(::Type{Ptr{T}}, u::TrixiMPIArray{T}) where {T} = Base.unsafe_convert(Ptr{T}, parent(u))
+Base.elsize(::Type{TrixiMPIArray{T, N, Parent}}) where {T, N, Parent} = elsize(Parent)
+
+
+# TODO: MPI. Do we need customized broadcasting? What about FastBroadcast.jl and
+#            threaded execution with `@.. thread=true`?
+# See https://docs.julialang.org/en/v1/manual/interfaces/#man-interfaces-broadcasting
+
+
+# Implementation of methods from ArrayInterface.jl for use with
+# LoopVectorization.jl etc.
+# See https://juliaarrays.github.io/ArrayInterface.jl/stable/
+ArrayInterface.parent_type(::Type{TrixiMPIArray{T, N, Parent}}) where {T, N, Parent} = Parent
+
+
+# TODO: MPI. Do we need LinearAlgebra methods such as `norm` or `dot`?
```

Comment:
Hm, difficult to say for me. On the one hand, it might make the interface nicer if those were supported. On the other hand, if you want to compute the `norm` over the entire array, the question is if the results really need to be available on all ranks (requiring a potentially expense all-to-all communication) or just on the root domain. Maybe do it like this: Do not implement it now, but once we touch code/write up something in the future that would benefit from such a functionality, let's revisit it.